map.c

```diff
@@ -409,8 +633,157 @@ typedef struct {
+	}
+	mm_time_seed_sum = mm_time_seed_avg;
+	mm_time_chain_sum = mm_time_chain_avg;
+	mm_time_align_sum = mm_time_align_avg;
+	mm_time_seed_avg /= p->n_threads;
+	mm_time_chain_avg /= p->n_threads;
+	mm_time_align_avg /= p->n_threads;
+}
+
+
+#if defined(__AMD_SPLIT_KERNELS__)
+static void worker_for(void *_data, long i, int tid) // kt_for() callback
+{
+	step_t *s = (step_t*)_data;
+	int j, iread, off, pe_ori = s->p->opt->pe_ori;
+	double t = 0.0;
+	mm_tbuf_t *b = s->buf[tid];
+	mm_trbuf_t *tr = s->trbuf[tid];
+
+	// Check if this is a valid read
+	if (i != -1) {
+
+		off = s->seg_off[i];
+		iread = tr->count;
+		tr->i[iread] = i;
+		tr->count++;
+
+		assert(s->n_seg[i] <= MM_MAX_SEG);
+		if (mm_dbg_flag & MM_DBG_PRINT_QNAME) {
+			fprintf(stderr, "QR\t%s\t%d\t%d\n", s->seq[off].name, tid, s->seq[off].l_seq);
+			t = realtime();
+		}
+
+		tr->qlens[iread] = (int *) malloc (s->n_seg[i] * sizeof(int));
+		tr->qseqs[iread] = (const char **) malloc (s->n_seg[i] * sizeof(const char *));
+		for (j = 0; j < s->n_seg[i]; ++j) {
+			if (s->n_seg[i] == 2 && ((j == 0 && (pe_ori>>1&1)) || (j == 1 && (pe_ori&1))))
+				mm_revcomp_bseq(&s->seq[off + j]);
+			tr->qlens[iread][j] = s->seq[off + j].l_seq;
+			tr->qseqs[iread][j] = s->seq[off + j].seq;
+		}
+		// Seed
+		if (s->p->opt->flag & MM_F_INDEPEND_SEG) {
+			for (j = 0; j < s->n_seg[i]; ++j) {
+				mm_map_seed(s->p->mi, 1, &tr->qlens[iread][j], &tr->qseqs[iread][j], &s->n_reg[off+j], &s->reg[off+j], b, s->p->opt, s->seq[off+j].name, &tr->rep_len[iread], &tr->qlen_sum[iread], &tr->n_mini_pos[iread], &tr->mini_pos[iread], &tr->n_a[iread], &tr->a[iread], tr->km[iread]);
+			}
+		} else {
+			mm_map_seed(s->p->mi, s->n_seg[i], tr->qlens[iread], tr->qseqs[iread], &s->n_reg[off], &s->reg[off], b, s->p->opt, s->seq[off].name, &tr->rep_len[iread], &tr->qlen_sum[iread], &tr->n_mini_pos[iread], &tr->mini_pos[iread], &tr->n_a[iread], &tr->a[iread], tr->km[iread]);
+		}
+	}
+
+	// Did we accumulate N_ACCUM reads or get to the last batch of reads?
+	if (tr->count == N_ACCUM || i == -1 ) {
+		// Chain
+		for (iread=0; iread<tr->count; iread++) {
+			i = tr->i[iread];
+			off = s->seg_off[i];
+			if (s->p->opt->flag & MM_F_INDEPEND_SEG) {
+				for (j = 0; j < s->n_seg[i]; ++j) {
+					mm_map_chain(s->p->mi, 1, &s->n_reg[off+j], &s->reg[off+j], b, s->p->opt, s->seq[off+j].name, &tr->rep_len[iread], &tr->frag_gap[iread], &tr->qlen_sum[iread], &tr->n_regs0[iread], &tr->n_mini_pos[iread], &tr->mini_pos[iread], &tr->n_a[iread], &tr->u[iread], &tr->a[iread], tr->km[iread]);
+					s->rep_len[off + j] = tr->rep_len[iread];
+					s->frag_gap[off + j] = tr->frag_gap[iread];
+				}
+			} else {
+				mm_map_chain(s->p->mi, s->n_seg[i], &s->n_reg[off], &s->reg[off], b, s->p->opt, s->seq[off].name, &tr->rep_len[iread], &tr->frag_gap[iread], &tr->qlen_sum[iread], &tr->n_regs0[iread], &tr->n_mini_pos[iread], &tr->mini_pos[iread], &tr->n_a[iread], &tr->u[iread], &tr->a[iread], tr->km[iread]);
```

Comment:
The backtracking still need to be done on CPU. But compact(a) relies on the result of backtracking. If we compact on CPU, we will have to copy anchors back from GPU. Therefore, we might copy chaining results back batch by batch (as our host memory cannot hold all of them), and add a non-blocking check on CPU side to run tasks everytime when a batch is copied back.

Reply:
When integrating the GPU kernel, you will call a different function (let's call it launch_gpu_chaining) instead of the for loop in line 727 above. Inside that function, you have the flexibility to work with the batch of seeding outputs given to you for tr->count reads. If backtracking on the CPU involves costly transfer of chaining output back to the host, then we may try to run backtracking on the GPU itself, perhaps in a single block of threads with 1 thread executing the backtracking algorithm. 
In the first iteration, you just wait for all the batches to be completed before sending control back out of this launch_gpu_chaining() function. We can brainstorm in our meeting.. there may be some things that I am missing.